Exception when generating RBI output

**Describe the bug**
I was testing out `sord-4.0.0` on one of my [YARDoced projects](https://github.com/ronin-rb/ronin-support/tree/1.0.0) (in the 1.0.0 branch) under ruby-3.1.2, but ran into this exception when attempting to generate RBI output.

**To Reproduce**
1. Add `gem 'sord'` to the `Gemfile`.
2. `bundle install`.
3. `bundle exec sord defs.rbi`

**Expected behavior**
Completes successfully.

**Actual behavior**

```
[OMIT ] (Ronin::Support::Binary::CTypes::Mixin#initialize_type_system) no YARD return type given, using untyped
[OMIT ] (Ronin::Support::Binary::CTypes::Mixin#initialize_type_system) no YARD return type given, using untyped
[ERROR] no implicit conversion of Parlour::Types::Raw into Integer
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/type_converter.rb:179:in `initialize'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/type_converter.rb:179:in `new'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/type_converter.rb:179:in `yard_to_parlour'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/type_converter.rb:116:in `block in yard_to_parlour'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/type_converter.rb:116:in `map'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/type_converter.rb:116:in `yard_to_parlour'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:364:in `block in add_methods'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:254:in `each'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:254:in `add_methods'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:553:in `add_namespace'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:558:in `block in add_namespace'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:558:in `each'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:558:in `add_namespace'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:558:in `block in add_namespace'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:558:in `each'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:558:in `add_namespace'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:558:in `block in add_namespace'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:558:in `each'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:558:in `add_namespace'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:570:in `block in populate'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:570:in `each'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:570:in `populate'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:590:in `run'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/parlour_plugin.rb:70:in `generate'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/exe/sord:88:in `block (2 levels) in <top (required)>'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/commander-4.6.0/lib/commander/command.rb:187:in `call'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/commander-4.6.0/lib/commander/command.rb:157:in `run'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/commander-4.6.0/lib/commander/runner.rb:444:in `run_active_command'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/commander-4.6.0/lib/commander/runner.rb:61:in `run!'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/commander-4.6.0/lib/commander/delegates.rb:18:in `run!'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/commander-4.6.0/lib/commander/import.rb:7:in `block in <top (required)>'
```

**Additional information**

* sord 4.0.0
* yard 0.9.28
* `ruby 3.1.2p20 (2022-04-12 revision 4491bb740a) [x86_64-linux]`

I get a similar error when attempting to generate RBS output, but it gets a little further:

```
[OMIT ] (Ronin::Support::Binary::CTypes::Mixin#initialize_type_system) no YARD return type given, using untyped
[OMIT ] (Ronin::Support::Binary::CTypes::Mixin#initialize_type_system) no YARD return type given, using untyped
[ERROR] no implicit conversion of Parlour::Types::Raw into Integer
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/type_converter.rb:179:in `initialize'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/type_converter.rb:179:in `new'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/type_converter.rb:179:in `yard_to_parlour'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/type_converter.rb:116:in `block in yard_to_parlour'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/type_converter.rb:116:in `map'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/type_converter.rb:116:in `yard_to_parlour'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:364:in `block in add_methods'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:254:in `each'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:254:in `add_methods'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:553:in `add_namespace'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:558:in `block in add_namespace'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:558:in `each'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:558:in `add_namespace'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:558:in `block in add_namespace'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:558:in `each'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:558:in `add_namespace'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:558:in `block in add_namespace'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:558:in `each'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:558:in `add_namespace'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:570:in `block in populate'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:570:in `each'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:570:in `populate'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/generator.rb:590:in `run'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/lib/sord/parlour_plugin.rb:70:in `generate'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/sord-4.0.0/exe/sord:88:in `block (2 levels) in <top (required)>'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/commander-4.6.0/lib/commander/command.rb:187:in `call'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/commander-4.6.0/lib/commander/command.rb:157:in `run'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/commander-4.6.0/lib/commander/runner.rb:444:in `run_active_command'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/commander-4.6.0/lib/commander/runner.rb:61:in `run!'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/commander-4.6.0/lib/commander/delegates.rb:18:in `run!'
         /data/home/postmodern/code/ronin-rb/vendor/bundle/ruby/3.1.0/gems/commander-4.6.0/lib/commander/import.rb:7:in `block in <top (required)>'
Parlour warning: RBS generation: 'type' is a keyword in RBS, renaming method parameter to '_type'
Parlour warning: RBS generation: 'type' is a keyword in RBS, renaming method parameter to '_type'
Parlour warning: RBS generation: 'type' is a keyword in RBS, renaming method parameter to '_type'
Parlour warning: RBS generation: 'type' is a keyword in RBS, renaming method parameter to '_type'
```

--- a/lib/sord/type_converter.rb
+++ b/lib/sord/type_converter.rb
@@ -160,23 +160,28 @@ module Sord
         generic_type = $1
         type_parameters = $2
 
+        # If we don't do this, `const_defined?` will resolve "::Array" as the actual Ruby `Array`
+        # type, not `Parlour::Types::Array`!
+        relative_generic_type = generic_type.start_with?('::') \
+          ? generic_type[2..] : generic_type
+
         parameters = split_type_parameters(type_parameters)
           .map { |x| yard_to_parlour(x, item, replace_errors_with_untyped, replace_unresolved_with_untyped) }
-        if SINGLE_ARG_GENERIC_TYPES.include?(generic_type) && parameters.length > 1
-          Parlour::Types.const_get(generic_type).new(Parlour::Types::Union.new(parameters))
-        elsif generic_type == 'Class' && parameters.length == 1
+        if SINGLE_ARG_GENERIC_TYPES.include?(relative_generic_type) && parameters.length > 1
+          Parlour::Types.const_get(relative_generic_type).new(Parlour::Types::Union.new(parameters))
+        elsif relative_generic_type == 'Class' && parameters.length == 1
           Parlour::Types::Class.new(parameters.first)
-        elsif generic_type == 'Hash'
+        elsif relative_generic_type == 'Hash'
           if parameters.length == 2
             Parlour::Types::Hash.new(*parameters)
           else
             handle_sord_error(parameters.map(&:describe).join, "Invalid hash, must have exactly two types: #{yard.inspect}.", item, replace_errors_with_untyped)
           end
         else
-          if Parlour::Types.const_defined?(generic_type)
+          if Parlour::Types.const_defined?(relative_generic_type)
             # This generic is built in to parlour, but sord doesn't
             # explicitly know about it.
-            Parlour::Types.const_get(generic_type).new(*parameters)
+            Parlour::Types.const_get(relative_generic_type).new(*parameters)
           else
             # This is a user defined generic
             Parlour::Types::Generic.new(

--- a/lib/sord/type_converter.rb
+++ b/lib/sord/type_converter.rb
@@ -163,7 +163,7 @@ module Sord
         # If we don't do this, `const_defined?` will resolve "::Array" as the actual Ruby `Array`
         # type, not `Parlour::Types::Array`!
         relative_generic_type = generic_type.start_with?('::') \
-          ? generic_type[2..] : generic_type
+          ? generic_type[2..-1] : generic_type
 
         parameters = split_type_parameters(type_parameters)
           .map { |x| yard_to_parlour(x, item, replace_errors_with_untyped, replace_unresolved_with_untyped) }

--- a/spec/type_converter_spec.rb
+++ b/spec/type_converter_spec.rb
@@ -216,6 +216,10 @@ describe Sord::TypeConverter do
           )
         end
       end
+
+      it 'does not resolve stdlib objects instead of types when using the root namespace' do
+        expect(subject.yard_to_parlour('::Array<String>')).to eq Types::Array.new('String')
+      end
     end
 
     context 'when given an untyped generic' do

